src/mainboard/starfive/visionfive2/bt0/src/main.rs

```diff
@@ -0,0 +1,297 @@
+#![feature(naked_functions, asm_const)]
+#![no_std]
+#![no_main]
+
+// #[macro_use]
+//extern crate log;
+// mod log;
+
+use core::{
+    arch::asm,
+    //    intrinsics::transmute,
+    panic::PanicInfo,
+    ptr::{read_volatile, write_volatile},
+    // slice,
+};
+use riscv::register::mhartid;
+// use riscv::register::{marchid, mimpid, mvendorid};
+// use embedded_hal::serial::nb::Write;
+// use init::{clock_init, iopad_init, rstgen_init};
+use uart::JH71XXSerial;
```

Comment:
Might be good to use the [uart_16550](https://docs.rs/uart_16550/latest/uart_16550/#with-memory-mapped-serial-port) crate here. Would work for both `visionfive1/2`, and other memory-mapped serial boards.

Reply:
Yea that one has its own hardcoded values though, and atomic instructions do not work on the JH7100 outside DRAM, i.e., this thing may go nuts in SRAM. That crate is rather for OSes than firmware.
See also:
https://github.com/rust-osdev/uart_16550/blob/d401ea103641cc06182c404840a5ba2896eebe72/src/mmio.rs#L52

We may just have our own internal crate instead, and I think we will. I am looking at multiple SoCs to figure this out. :-)

Reply:
Makes sense, didn't know if there was enough crossover to be usable in firmware.

Also didn't know about atomics not working in SRAM...

> We may just have our own internal crate instead

That would be great!

Reply:
Just opened a pull request to allow custom UART init in [uart_16550](https://github.com/rust-osdev/uart_16550/pull/25).

If we can solve the issue with atomics being wonky in SRAM, maybe we can use their lib in `oreboot`?

Could also try replacing their atomic use with a `SpinLock`, or a wrapper that conditionally enables memory barriers with a feature. That way we could disable for the case of the bootloader running on a single core, single thread.

Reply:
Do you have a VF2 to experiment with? Would be nice if you could try this out. I'm currently translating the DRAM init, which takes quite a while...

Reply:
> Do you have a VF2 to experiment with?

Yes, I'll try flashing what you have now. Is the current stuff you have bootable?

Also, I was going to start working on a QEMU setup for testing. Does the current emulation board work for VF2?

Reply:
The current code is neither for flashing nor QEMU.

You can go back some commits to have something that can be run from UART. Note that there is no clock/PLL setup yet, so we just pick up what the mask ROM did.
See also the notes in the Makefile.

Reply:
The cureent QEMU board is linked at an address that differs from the JH7110's SRAM. And it uses the ns16550 crate, which is unlikely to work. Feel free to experiment though.

Remember to wrap the binary in the necessary header.
I recommend the UART loader instead of flashing for convenience.
See also the recordings for reference:
https://youtube.com/playlist?list=PLenOHeTI_A9MJlYIOAVC0JDpKKXX9mZgK

Reply:
> The cureent QEMU board is linked at an address that differs from the JH7110's SRAM. And it uses the ns16550 crate, which is unlikely to work.

Yeah, I started looking at the QEMU board, and docs on the main QEMU site. I didn't know QEMU was so much more specific for ARM and RISCV boards (opposed to x86-64). Makes sense with ARM/RISCV being more embedded boards with wildly different peripherals and memory setup.

> I recommend the UART loader instead of flashing for convenience.

Awesome, I'll check out the linked video, and try to see how far I get.